### PR TITLE
Add metrics for number of segments generated per task in MSQ

### PR DESCRIPTION
**Pull request by YongGang**
### Description

Report `ingest/tombstones/count` and `ingest/segments/count` metrics in MSQ when finish publishing segments.

Following are two examples:
```
{"taskIngestionMode":"NONE","feed":"metrics","taskType":"query_controller","metric":"ingest/tombstones/count","service":"druid/indexer","groupId":"query-481a0c2d-1f04-4a5b-bc30-f343240850d3","host":"localhost:8091","version":"","value":0,"dataSource":"wikipedia","taskId":"query-481a0c2d-1f04-4a5b-bc30-f343240850d3","timestamp":"2023-09-14T17:01:47.951Z"}

{"taskIngestionMode":"NONE","feed":"metrics","taskType":"query_controller","metric":"ingest/segments/count","service":"druid/indexer","groupId":"query-481a0c2d-1f04-4a5b-bc30-f343240850d3","host":"localhost:8091","version":"","value":1,"dataSource":"wikipedia","taskId":"query-481a0c2d-1f04-4a5b-bc30-f343240850d3","timestamp":"2023-09-14T17:01:47.954Z"}
```

#### Release note

Add `ingest/tombstones/count` and `ingest/segments/count` metrics in MSQ.

<hr>

##### Key changed/added classes in this PR
 * In `ControllerImpl` report the new metrics

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Review by LakshSingla — COMMENTED** [review]
1. Is there a way to distinguish between the metrics generated by MSQ and the native ingestion? Since we are selectively adding a few metrics, it would be nice to be able to filter by the mode of ingestion.
2. We'd wanna document this in the MSQ docs and/or the metrics docs somehow

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java [hunk 1]
```diff
@@ -1344,6 +1345,7 @@ private void publishAllSegments(final Set<DataSegment> segments) throws IOExcept
               destination.getSegmentGranularity()
           );
           segmentsWithTombstones.addAll(tombstones);
+          tombstoneSize = tombstones.size();
```
Comment (kfaraz) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java:1348]:
```suggestion
          numTombstones = tombstones.size();
```

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java [hunk 2]
```diff
@@ -1389,6 +1391,10 @@ private void publishAllSegments(final Set<DataSegment> segments) throws IOExcept
           SegmentTransactionalInsertAction.appendAction(segments, null, null)
       );
     }
+
+    task.emitMetric(context.emitter(), "ingest/tombstones/count", tombstoneSize);
+    // segments count metric is documented to include tombstones
```
Comment (kfaraz) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java:1396]:
```suggestion
    // Include tombstones in the reported segments count
```

File: extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/test/MSQTestControllerContext.java [hunk 3]
```diff
@@ -215,6 +216,12 @@ public void close()
     }
   };
 
+  @Override
+  public ServiceEmitter emitter()
+  {
+    return null;
```
[the remark this answers is not in the document]
Comment (kfaraz) [reply extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/test/MSQTestControllerContext.java:222]: Or you can also use `StubServiceEmitter` too if you intend to capture and verify metrics in the tests. Also, maybe instantiate this object just once and not on every invocation of `emitter`.

**Comment by kfaraz** [comment]
> Is there a way to distinguish between the metrics generated by MSQ and the native ingestion? Since we are selectively adding a few metrics, it would be nice to be able to filter by the mode of ingestion.

I suppose you could filter by the type of task that has reported that metric.

**Comment by YongGang** [comment]
> > Is there a way to distinguish between the metrics generated by MSQ and the native ingestion? Since we are selectively adding a few metrics, it would be nice to be able to filter by the mode of ingestion.
> 
> @LakshSingla , I suppose you could filter by the type of task that has reported that metric.

yup, can use task type to distinguish. I added two examples of the metrics generated in the PR description.

**Review by kfaraz — APPROVED** [review]
LGTM

**Comment by YongGang** [comment]
Hi @LakshSingla , does this change look good to you? The main purpose for these two metrics is for alerting: when a task creates way many segments, it will make the cluster unstable, we want to be alerted in this case.

**Comment by YongGang** [comment]
fyi @cryptoe

**Comment by LakshSingla** [comment]
Merging the PR, cc @cryptoe if we require doc changes for the same.

**Pull request merged**